File: LC_JS-SOLUTIONS/39. Combination Sum.js
/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
var combinationSum = function(candidates, target) {


    function backTrack(start,target,path){

        if (target == 0){
            result.push([...path])
            return
        }

        for(let i = start; i<candidates.length; i++){

            if (candidates[i]>target){
                continue
            }

            path.push(candidates[i])
            backTrack(i, target-candidates[i],path)
            path.pop()

        }

    }



    let result = []
    candidates.sort((a,b)=>a-b)
    backTrack(0,target,[])
    return result



    
};

File: LC-Python-Solution/39. Combination Sum.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:

        def backTrack(start, target, path):
            if target == 0:
                result.append(path.copy())
                return
            
            for i in range(start,len(candidates)):
                if candidates[i] > target:
                    continue
                
                path.append(candidates[i])
                backTrack(i,target-candidates[i],path)
                path.pop()


        result = []
        candidates.sort()
        backTrack(0,target,[])
        return result

        


        